pandas/tests/test_categorical.py

```diff
@@ -4549,6 +4549,13 @@ def test_concat_categorical(self):
                             'h': [None] * 6 + cat_values})
         tm.assert_frame_equal(res, exp)
 
+    def test_rank_categorical(self):
+        exp = pd.Series([1., 2., 3., 4., 5., 6.], name='A')
+        dframe = pd.DataFrame(['first', 'second', 'third', 'fourth', 'fifth', 'sixth'], columns=['A'])
+        dframe['A'] = dframe['A'].astype('category', ).cat.set_categories(
+                ['first', 'second', 'third', 'fourth', 'fifth', 'sixth'], ordered=True)
+        res = dframe['A'].rank()
+        tm.assert_series_equal(res, exp)
```

Comment:
I actually think it might be best to simply define ``.rank()`` on a ``Categorical``; here you would just check if ``._values`` has a ``.rank()`` method, if so, then defer to it and you are done.

Inside ``Categorical`` you can THEN construct an object you want ranked (a Series) with the appropriate value (e.g. the codes), but you will have to substitute NaN if any -1's. Then just pass thru the resulting rankings.